radix-engine/src/updates/protocol_updates.rs

```diff
@@ -0,0 +1,106 @@
+use super::state_updates::*;
+use crate::{internal_prelude::*, track::StateUpdates};
+use radix_substate_store_interface::interface::SubstateDatabase;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ProtocolUpdate {
+    /// Introduces BLS12-381 and Keccak-256 features.
+    Bls12381AndKeccak256,
+
+    /// Exposes second-precision timestamp.
+    SecondPrecisionTimestamp,
+
+    /// Increases the math precision with native pool implementations.
+    PoolMathPrecisionFix,
+
+    /// Changes the cost associated with validator creation.
+    ValidatorCreationFeeFix,
+
+    /// Exposes a getter method for reading owner role rule.
+    OwnerRoleGetter,
+
+    /// Various system patches.
+    SystemPatches,
+
+    /// Introduces the account locker blueprint.
+    AccountLocker,
+
+    /// Moves various protocol parameters to state.
+    ProtocolParamsToState,
+}
+
+pub struct ProtocolUpdates {
+    updates: BTreeSet<ProtocolUpdate>,
+}
+
+impl ProtocolUpdates {
+    pub fn none() -> Self {
+        Self {
+            updates: BTreeSet::new(),
+        }
+    }
+
+    pub fn all() -> Self {
+        Self::none().with_anemone().with_bottlenose()
+    }
+
+    /// Enables all the protocol updates included in the `anemone` release.
+    pub fn with_anemone(mut self) -> Self {
+        self.updates.extend(btreeset!(
+            ProtocolUpdate::Bls12381AndKeccak256,
+            ProtocolUpdate::SecondPrecisionTimestamp,
+            ProtocolUpdate::PoolMathPrecisionFix,
+            ProtocolUpdate::ValidatorCreationFeeFix,
+        ));
+        self
+    }
+
+    /// Enables all the protocol updates included in the `bottlenose` release.
+    ///
+    /// Note that this does not include `anemone` protocol updates.
+    pub fn with_bottlenose(mut self) -> Self {
+        self.updates.extend(btreeset!(
+            ProtocolUpdate::OwnerRoleGetter,
+            ProtocolUpdate::SystemPatches,
+            ProtocolUpdate::AccountLocker,
+            ProtocolUpdate::ProtocolParamsToState,
+        ));
+        self
+    }
+
+    pub fn with_update(mut self, update: ProtocolUpdate) -> Self {
+        self.updates.insert(update);
+        self
+    }
+
+    pub fn without_update(mut self, update: ProtocolUpdate) -> Self {
+        self.updates.remove(&update);
+        self
+    }
+
+    pub fn generate_state_updates<S: SubstateDatabase>(&self, db: &S) -> Vec<StateUpdates> {
```

Comment:
Perhaps the state updates should be provided in a canonical order and not in the order in which the enum variants are provided?

Reply:
Ah yes you're right. In this case we should remember not to change the ordering of the variants on the enum.